src/lib.rs

```diff
@@ -180,8 +180,23 @@ mod contract {
     /// code.
     #[no_mangle]
     pub extern "C" fn state_migration() {
-        // This function is purposely left empty because we do not have any state migration
-        // to do.
+        // Only owner can call migration
+        sdk::assert_private_call();
+
+        let metadata_key = bytes_to_key(
+            KeyPrefix::EthConnector,
+            &[EthConnectorStorageId::FungibleTokenMetadata as u8],
+        );
+
+        //=========================================================
+        // Migrate Metadata
+        if !sdk::storage_has_key(&metadata_key[..]) {
+            use crate::fungible_token::FungibleTokenMetadata;
+
+            let metadata =
+                FungibleTokenMetadata::try_from_slice(&sdk::read_input()).expect(ERR_FAILED_PARSE);
```

Comment:
The `state_migration` function is not called directly by a user, it is [called as part of the upgrade process](https://github.com/aurora-is-near/aurora-engine/blob/develop/src/sdk.rs#L456-L462), so you would need to thread this data through from the [`deploy_upgrade` function](https://github.com/aurora-is-near/aurora-engine/blob/develop/src/lib.rs#L169).

To see an example of the upgrade flow (stage -> deploy -> state_migration), we have the [state migration test](https://github.com/aurora-is-near/aurora-engine/blob/develop/src/tests/state_migration.rs#L12).

Reply:
Looks as really bad design for migration.
You can't control migration, for that design.

Reply:
We should put parameters to `sefl_upgrade` all call migration directly.
But for all the best migrations that I know - it should be a direct call. Otherwise, you can't control it.

Reply:
It was suggested we do it this way in [Eugene Kuzyakov's review](https://github.com/aurora-is-near/aurora-engine/pull/74/files#diff-8bcc54eddebf73a8eb83cf6739e4e992583f4fb2cee5c7a8f317680b465cfdfbR333).

His argument was that upgrade should include state migration otherwise you could end up in a broken place where the new code doesn't understand how to read the old state. That's why we have state migration automatically called during upgrade.